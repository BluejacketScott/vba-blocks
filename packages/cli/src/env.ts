import { join } from 'path';
import { homedir } from 'os';
import { Reporter, reporter } from './reporter';

export interface Env {
  isWindows: boolean;
  cwd: string;
  values: any;

  addins: string;
  scripts: string;
  cache: string;
  registry: string;
  packages: string;
  sources: string;
  staging: string | null;

  reporter: Reporter;
}

const cache = join(homedir(), '.vba-blocks');
const env: Env = {
  isWindows: process.platform === 'win32',
  cwd: process.cwd(),
  values: process.env,

  addins: join(__dirname, '../../addin/build'),
  scripts: join(__dirname, '../scripts'),
  cache,
  registry: join(cache, 'registry'),
  packages: join(cache, 'packages'),
  sources: join(cache, 'sources'),
  staging: null,

  reporter
};

export default env;
