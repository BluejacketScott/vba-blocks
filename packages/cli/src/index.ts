export { default as env } from './env';
export {
  Version,
  Dependency,
  Feature,
  Reference,
  Source,
  Target,
  Manifest,
  loadManifest
} from './manifest';
export { Config, loadConfig } from './config';
export { default as resolve } from './resolve';
