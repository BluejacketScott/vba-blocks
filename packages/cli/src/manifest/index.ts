import { join } from 'path';
import { ok } from 'assert';
import { parse as parseToml } from 'toml';
import { pathExists, readFile } from '../utils/fs';
import { Version } from './version';
import { Source, parseSrc } from './source';
import { Feature, parseFeatures } from './feature';
import { Dependency, parseDependencies } from './dependency';
import { Reference, parseReferences } from './reference';
import { Target, parseTargets } from './target';

export { Version, Source, Feature, Dependency, Reference, Target };

/**
 * @example
 * ```toml
 * [package]
 * name = "package-name"
 * version = "1.0.0-rc.1"
 * authors = ["Tim Hall <tim.hall.engr@gmail.com> (https://github.com/timhall)"]
 * 
 * [src]
 * A = "src/a.bas"
 * B = { path = "src/b.cls" }
 * C = { path = "src/c.frm", optional = true }
 * 
 * [features]
 * default = ["a"]
 * 
 * a = { src = ["C"] }
 * b = { dependencies = ["dictionary"] }
 * c = { references = ["Scripting"] }
 * 
 * [dependencies]
 * dictionary = "v1.4.1"
 * with-features = { version = "1.0.0", default-features = false, features = ["other"] }
 * from-path = { path = "packages/from-path" }
 * from-git-master = { git = "https://github.com/VBA-tools/VBA-Web.git" }
 * from-git-branch = { git = "https://github.com/VBA-tools/VBA-Web.git", branch = "beta" }
 * from-git-tag = { git = "https://github.com/VBA-tools/VBA-Web.git", tag = "v1.0.0" }
 * from-git-rev = { git = "https://github.com/VBA-tools/VBA-Web.git", rev = "a1b2c3d4" }
 * 
 * [references.Scripting]
 * version = "1.0"
 * guid = "{420B2830-E718-11CF-893D-00A0C9054228}"
 * optional = true
 * 
 * [[targets]]
 * type = "xlsm"
 * path = "targets/xlsm"
 * 
 * [[targets]]
 * name = "custom-name"
 * type = "xlam"
 * path = "targets/xlam"
 * ```
 */

export interface Snapshot {
  name: string;
  version: Version;
  dependencies: Dependency[];
}

export interface Metadata {
  authors: string[];
  publish: boolean;
  [name: string]: any;
}

export interface Manifest extends Snapshot {
  metadata: Metadata;
  src: Source[];
  references: Reference[];
  targets: Target[];
  features: Feature[];
  defaultFeatures: string[];
  dir: string;
}

const EXAMPLE = `Example vba-block.toml:

  [package]
  name = "my-package"
  version = "0.0.0"
  authors = ["..."]`;

export function parseManifest(value: any, dir: string): Manifest {
  ok(
    value && value.package,
    `[package] is a required field, with name, version, and authors specified. ${EXAMPLE}`
  );

  const { name, version, authors, publish = false } = value.package;

  ok(name, `[package] name is a required field. ${EXAMPLE}`);
  ok(version, `[package] version is a required field. ${EXAMPLE}`);
  ok(authors, `[package] authors is a required field. ${EXAMPLE}`);

  const src = parseSrc(value.src || {}, dir);
  const { features, defaultFeatures } = parseFeatures(value.features || {});
  const dependencies = parseDependencies(value.dependencies || {});
  const references = parseReferences(value.references || {});
  const targets = parseTargets(value.targets || [], name, dir);

  const metadata = { publish, ...value.package };

  return {
    name,
    version,
    metadata,
    src,
    features,
    defaultFeatures,
    dependencies,
    references,
    targets,
    dir
  };
}

export async function loadManifest(dir: string): Promise<Manifest> {
  const file = join(dir, 'vba-block.toml');
  if (!await pathExists(file)) {
    throw new Error(`vba-blocks.toml not found in "${dir}"`);
  }

  const raw = await readFile(file);
  const parsed = parseToml(raw.toString());
  const manifest = parseManifest(parsed, dir);

  return manifest;
}
