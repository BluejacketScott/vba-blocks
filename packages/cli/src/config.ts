import { join } from 'path';
import { parse as parseToml } from 'toml';
import env from './env';
import { pathExists, readFile } from './utils';
import {
  Sources,
  RegistrySource,
  PathSource,
  GitSource,
  UnsupportedSource
} from './sources';

export type Registry =
  | {}
  | { [name: string]: { index: string; packages: string } };

export interface Flags {
  git?: boolean;
  path?: boolean;
}

export interface Config {
  registry: Registry;
  flags: Flags;
  sources: Sources;
}

export interface ConfigValue {
  registry?: Registry;
  flags?: Flags;
}

const empty: ConfigValue = { registry: {}, flags: {} };
const defaults: ConfigValue = {
  registry: {
    'vba-blocks': {
      index: 'https://github.com/vba-blocks/registry',
      packages: 'https://packages.vba-blocks.com'
    }
  },
  flags: { git: true, path: true }
};

/**
 * Load config, from local, user, and environment values
 *
 * - Search for .vba-blocks/config.toml up from cwd
 * - Load ~/.vba-blocks/config.toml
 * - Load VBA_BLOCKS_* from environment
 */
export async function loadConfig(): Promise<Config> {
  const user: ConfigValue = {
    ...empty,
    ...((await readConfig(env.cache)) || {})
  };
  const file = await findConfig(env.cwd);
  const local: ConfigValue = {
    ...empty,
    ...file ? await readConfig(file) : {}
  };
  const override = loadConfigFromEnv();

  const registry: Registry = {
    ...defaults.registry,
    ...user.registry,
    ...local.registry,
    ...override.registry
  };
  const flags: Flags = {
    ...defaults.flags,
    ...user.flags,
    ...local.flags,
    ...override.flags
  };

  const sources: Sources = {
    registry: {},
    git: flags.git ? new GitSource() : new UnsupportedSource('git'),
    path: flags.path ? new PathSource() : new UnsupportedSource('path')
  };

  for (const [name, { index, packages }] of Object.entries(registry)) {
    sources.registry[name] = new RegistrySource({
      name,
      index,
      packages
    });
  }

  return { registry, flags, sources };
}

// Read config from dir (if present)
export async function readConfig(
  dir: string
): Promise<ConfigValue | undefined> {
  const file = join(dir, 'config.toml');
  if (!await pathExists(file)) return {};

  const raw = await readFile(file);
  const parsed = parseToml(raw.toString());

  return parsed;
}

// Find config up from and including given dir
// (looking for .vba-blocks/config.toml)
export async function findConfig(dir: string): Promise<string | undefined> {
  // TODO Search from .vba-blocks/config.toml starting at cwd
  return;
}

// Load config values from environment
export function loadConfigFromEnv(): ConfigValue {
  // TODO Load override config from VBA_BLOCKS_* env variables
  return {};
}
