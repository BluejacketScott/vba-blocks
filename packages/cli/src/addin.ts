import env from './env';
import { Project } from './project';
import { Target, Source, Reference } from './manifest';
import { run, unixJoin } from './utils';

export type Application = string;
export type Addin = string;

export const extensions: { [application: string]: string[] } = {
  excel: ['xlsx', 'xlsm', 'xlam']
};
export const addins: { [application: string]: string } = {
  excel: 'vba-blocks.xlsm' // TEMP vba-blocks.xlam
};

const byExtension: { [extension: string]: string } = {};
for (const [application, values] of Object.entries(extensions)) {
  for (const extension of values) {
    byExtension[extension] = application;
  }
}

/**
 * Import graph of src and references into given target
 */
export async function importGraph(
  project: Project,
  target: Target,
  graph: { src: Source[]; references: Reference[] }
): Promise<void> {
  const { application, addin, file } = getTargetInfo(project, target);
  const { src, references } = graph;

  await run(application, addin, 'Build.ImportGraph', {
    file,
    src,
    references
  });
}

/**
 * Export src and references from given target
 */
export async function exportTo(
  project: Project,
  target: Target,
  staging: string
): Promise<void> {
  const { application, addin, file } = getTargetInfo(project, target);

  const result = await run(application, addin, 'Build.ExportTo', {
    file,
    staging
  });
  console.log('export result', result);
}

/**
 * Get application, addin, and file for given target
 */
export function getTargetInfo(
  project: Project,
  target: Target
): { application: Application; addin: Addin; file: string } {
  const application = byExtension[target.type];
  if (!application) throw new Error(`Unsupported target type "${target.type}"`);

  const addin = unixJoin(env.addins, addins[application]);
  const file = unixJoin(project.paths.build, target.filename);

  return { application, addin, file };
}
