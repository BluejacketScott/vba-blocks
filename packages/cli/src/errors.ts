import chalk from 'chalk';
import env from './env';
import { cleanError, isString } from './utils';

import { Target } from './manifest';
import { Registration } from './sources';

export interface CliErrors {
  'unknown-command': (command: string) => string;

  'unsupported-source': (type: string) => string;
  'dependency-not-found': (dependency: string, registry: string) => string;
  'dependency-invalid-checksum': (registration: Registration) => string;

  'target-not-found': (target: Target) => string;
  'target-is-open': (target: Target, path: string) => string;
  'target-import-failed': (target: Target) => string;
  'target-restore-failed': (backup: string, file: string) => string;

  'resolve-failed': (details?: string) => string;

  'run-script-not-found': (path: string) => string;
}

export interface CliErrorOptions {
  code?: string;
  underlying?: Error;
}

export class CliError extends Error {
  code?: string;
  underlying?: Error;

  constructor(message: string, options: CliErrorOptions = {}) {
    if (options.underlying) {
      const underlying = cleanError(options.underlying);
      message += chalk`\n\n{dim ${underlying.message}}`;
    }

    super(message);

    this.code = options.code;
    this.underlying = options.underlying;
  }
}

// TODO Resolve typing issues so the following isn't so repetitive/verbose

export function unknownCommand(command: string): Error {
  const code = 'unknown-command';
  const message = env.reporter.errors[code](command);

  return new CliError(message, { code });
}

export function unsupportedSource(type: string): Error {
  const code = 'unsupported-source';
  const message = env.reporter.errors[code](type);

  return new CliError(message, { code });
}

export function dependencyNotFound(
  dependency: string,
  registry: string
): Error {
  const code = 'dependency-not-found';
  const message = env.reporter.errors[code](dependency, registry);

  return new CliError(message, { code });
}

export function dependencyInvalidChecksum(registration: Registration): Error {
  const code = 'dependency-invalid-checksum';
  const message = env.reporter.errors[code](registration);

  return new CliError(message, { code });
}

export function targetNotFound(target: Target): Error {
  const code = 'target-not-found';
  const message = env.reporter.errors[code](target);

  return new CliError(message, { code });
}

export function targetIsOpen(target: Target, path: string): Error {
  const code = 'target-is-open';
  const message = env.reporter.errors[code](target, path);

  return new CliError(message, { code });
}

export function targetImportFailed(target: Target, underlying: Error): Error {
  const code = 'target-import-failed';
  const message = env.reporter.errors[code](target);

  return new CliError(message, { code, underlying });
}

export function targetRestoreFailed(
  backup: string,
  file: string,
  underlying: Error
): Error {
  const code = 'target-restore-failed';
  const message = env.reporter.errors[code](backup, file);

  return new CliError(message, { code, underlying });
}

export function resolveFailed(details?: string): Error {
  const code = 'resolve-failed';
  let message = env.reporter.errors[code]();
  if (details) {
    message += `\n${details}`;
  }

  return new CliError(message, { code });
}

export function runScriptNotFound(path: string): Error {
  const code = 'run-script-not-found';
  const message = env.reporter.errors[code](path);

  return new CliError(message, { code });
}
