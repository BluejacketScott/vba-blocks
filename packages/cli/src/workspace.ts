import { Manifest } from './manifest';

export interface Workspace {
  root: Manifest;
  members: Manifest[];
}

export async function loadWorkspace(manifest: Manifest): Promise<Workspace> {
  // TODO Load workspace
  return {
    root: manifest,
    members: []
  };
}
