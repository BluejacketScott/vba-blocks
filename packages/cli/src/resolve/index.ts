import { Config } from '../config';
import { Workspace } from '../workspace';
import { Project } from '../project';
import { DependencyGraph, fetchDependencies } from './dependency-graph';
import { readLockfile, isLockfileValid } from '../lockfile';
import Resolver from './resolver';
import solveLatest from './latest-solver';
import solveSat from './sat-solver';

export { DependencyGraph, fetchDependencies, Resolver, solveLatest, solveSat };

export default async function resolve(
  config: Config,
  workspace: Workspace
): Promise<DependencyGraph> {
  const lockfile = await readLockfile(config);

  // If lockfile is up-to-date with manifest, use cached dependency graph
  if (lockfile && !isLockfileValid(lockfile, workspace)) {
    return lockfile.packages;
  }

  // Load and update resolver
  const resolver = new Resolver(config);
  await resolver.update();

  // Seed preferred versions on resolver
  const preferred = lockfile ? lockfile.packages : [];
  resolver.prefer(preferred);

  // Attempt latest solver, saving errors for recommendations on failure
  let conficts;
  try {
    return await solveLatest(config, workspace, resolver);
  } catch (err) {
    // TODO extract conflicts from latest solver error
  }

  // Fallback to SAT solver
  try {
    return await solveSat(config, workspace, resolver);
  } catch (err) {
    // No useful information from SAT solver failure, ignore
  }

  // TODO Include conflicts with error
  throw new Error('Unable to resolve dependency graph for given manifest');
}
