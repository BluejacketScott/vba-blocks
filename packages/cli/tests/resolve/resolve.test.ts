import resolve from '../../src/resolve';
import { loadConfig } from '../../src/config';
import { toWorkspace } from '../helpers/workspace';
import * as manifest from '../fixtures/manifest';

jest.mock('../../src/sources/registry-source');

test('solves simple tree', async () => {
  const config = await loadConfig();
  const solution = await resolve(config, toWorkspace(manifest.simple));
  expect(solution).toMatchSnapshot();
});

test('solves complex tree', async () => {
  const config = await loadConfig();
  const solution = await resolve(config, toWorkspace(manifest.complex));
  expect(solution).toMatchSnapshot();
});

test('solves needs-sat tree', async () => {
  const config = await loadConfig();
  const solution = await resolve(config, toWorkspace(manifest.needsSat));
  expect(solution).toMatchSnapshot();
});

test('fails to solve unresolvable tree', async () => {
  const config = await loadConfig();
  await expect(() => resolve(config, toWorkspace(manifest.unresolvable)))
    .rejects;
});
