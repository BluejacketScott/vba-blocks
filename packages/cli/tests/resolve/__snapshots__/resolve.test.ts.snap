// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`fails to solve unresolvable tree 1`] = `[Error: Unable to resolve dependency graph for given manifest]`;

exports[`solves complex tree 1`] = `
Array [
  Object {
    "dependencies": Array [
      Object {
        "name": "d",
        "registry": "vba-blocks",
        "version": "^2",
      },
    ],
    "id": "a@1.2.0",
    "name": "a",
    "source": "registry+vba-blocks#<hash>",
    "version": "1.2.0",
  },
  Object {
    "dependencies": Array [
      Object {
        "name": "c",
        "registry": "vba-blocks",
        "version": "^0.1.5",
      },
    ],
    "id": "b@1.1.0",
    "name": "b",
    "source": "registry+vba-blocks#<hash>",
    "version": "1.1.0",
  },
  Object {
    "dependencies": Array [
      Object {
        "name": "f",
        "registry": "vba-blocks",
        "version": "^2",
      },
    ],
    "id": "d@2.0.0",
    "name": "d",
    "source": "registry+vba-blocks#<hash>",
    "version": "2.0.0",
  },
  Object {
    "dependencies": Array [
      Object {
        "name": "g",
        "registry": "vba-blocks",
        "version": "^1",
      },
    ],
    "id": "f@2.0.0",
    "name": "f",
    "source": "registry+vba-blocks#<hash>",
    "version": "2.0.0",
  },
  Object {
    "dependencies": Array [],
    "id": "g@1.0.0",
    "name": "g",
    "source": "registry+vba-blocks#<hash>",
    "version": "1.0.0",
  },
  Object {
    "dependencies": Array [],
    "id": "c@0.1.6",
    "name": "c",
    "source": "registry+vba-blocks#<hash>",
    "version": "0.1.6",
  },
]
`;

exports[`solves needs-sat tree 1`] = `
Array [
  Object {
    "dependencies": Array [
      Object {
        "name": "d",
        "registry": "vba-blocks",
        "version": "^1.2.0",
      },
    ],
    "id": "a@1.1.0",
    "name": "a",
    "source": "registry+vba-blocks#<hash>",
    "version": "1.1.0",
  },
  Object {
    "dependencies": Array [
      Object {
        "name": "c",
        "registry": "vba-blocks",
        "version": "^0.1.0",
      },
    ],
    "id": "b@1.0.0",
    "name": "b",
    "source": "registry+vba-blocks#<hash>",
    "version": "1.0.0",
  },
  Object {
    "dependencies": Array [],
    "id": "c@0.1.0",
    "name": "c",
    "source": "registry+vba-blocks#<hash>",
    "version": "0.1.0",
  },
  Object {
    "dependencies": Array [
      Object {
        "name": "f",
        "registry": "vba-blocks",
        "version": "^1",
      },
    ],
    "id": "d@1.2.0",
    "name": "d",
    "source": "registry+vba-blocks#<hash>",
    "version": "1.2.0",
  },
  Object {
    "dependencies": Array [
      Object {
        "name": "g",
        "registry": "vba-blocks",
        "version": "^1",
      },
    ],
    "id": "f@1.0.0",
    "name": "f",
    "source": "registry+vba-blocks#<hash>",
    "version": "1.0.0",
  },
  Object {
    "dependencies": Array [],
    "id": "g@1.0.0",
    "name": "g",
    "source": "registry+vba-blocks#<hash>",
    "version": "1.0.0",
  },
]
`;

exports[`solves simple tree 1`] = `
Array [
  Object {
    "dependencies": Array [
      Object {
        "name": "d",
        "registry": "vba-blocks",
        "version": "^2",
      },
    ],
    "id": "a@1.2.0",
    "name": "a",
    "source": "registry+vba-blocks#<hash>",
    "version": "1.2.0",
  },
  Object {
    "dependencies": Array [
      Object {
        "name": "f",
        "registry": "vba-blocks",
        "version": "^2",
      },
    ],
    "id": "d@2.0.0",
    "name": "d",
    "source": "registry+vba-blocks#<hash>",
    "version": "2.0.0",
  },
  Object {
    "dependencies": Array [
      Object {
        "name": "g",
        "registry": "vba-blocks",
        "version": "^1",
      },
    ],
    "id": "f@2.0.0",
    "name": "f",
    "source": "registry+vba-blocks#<hash>",
    "version": "2.0.0",
  },
  Object {
    "dependencies": Array [],
    "id": "g@1.0.0",
    "name": "g",
    "source": "registry+vba-blocks#<hash>",
    "version": "1.0.0",
  },
]
`;
